### ISSUE TITLE
Threading client raise error and stop program if subscription PV is not found

### ISSUE THREAD

**Issue by vstadnytskyi** [caproto/caproto#539]
I have a ThreadingClient that subscribes to a set of PVs. however, it raises an error and crashes if I am trying to subscribe do not exist yet. The raised error stops the execution of the rest of the code. So the problem is, I cannot lunch a threading client if servers are not running. Am I doing something wrong? 

``` ipython3
$ ipython3
Python 3.7.4 (v3.7.4:e09359112e, Jul  8 2019, 14:54:52)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.8.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: pv_name = 'test'
In [2]: from caproto.threading.client import Context
In [3]: ctx = Context()
In [4]: pv,  = ctx.get_pvs(pv_name)
In [5]: sub = pv.subscribe()
In [6]: def func(reponse):
   ...:     print(response)
In [7]: sub.add_callback(func)
---------------------------------------------------------------------------
CaprotoTimeoutError                       Traceback (most recent call last)
<ipython-input-7-e3673ad5ef5d> in <module>
----> 1 sub.add_callback(func)

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/caproto/threading/client.py in add_callback(self, func)
   2176             # current value to this func (and any other funcs added in the
   2177             # mean time).
-> 2178             self._subscribe()
   2179         else:
   2180             # This callback is piggy-backing onto an existing subscription.

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/caproto/threading/client.py in inner(self, *args, **kwargs)
    103                 if not ready:
    104                     raise CaprotoTimeoutError(
--> 105                         f"{pv} could not connect within "
    106                         f"{float(raw_timeout):.3}-second timeout.")
    107                 if timeout is not None:

CaprotoTimeoutError: <PV name='test' priority=0 (searching....)> could not connect within 2.0-second timeout.
```

**Comment by danielballan** [comment]
This is expected behavior. We interpret adding a callback to a subscription as saying, "Start monitoring this PV now," the same way we would interpret `pv.read()` as saying, "Read this PV now." If the PV can't be connected to, we raise an error to let you know that you aren't collecting that readings that (we assume) you are expecting.

You could get away with something like:

```py
def add_callback_ignore_timeout(sub, func):
     try:
         sub.add_callback(func)
     except caproto.CaprotoTimeoutError:
        pass
```

but I think it might be better to turn the server on first.

**Comment by vstadnytskyi** [comment]
>... you aren't collecting that readings that (we assume) you are expecting.

That is correct I am expecting them and if there is nothing to read I should be notified. and iot is OK to raise an error in case of read() because I can catch it and keep going. However, in the case of subscribe() it just crashes the program. This is behavior is unexpected for me. 

For example,  PyEpics GUI can start if the server is running and it does not crash when PV is not available it just stops updating. And when the server is back up running it starts updating. This is expected. 

I have run some tests, to see what will happen if the server is rebooted after the client has subscribed. Nothing happened. The client kept running. And the client's subscription was still alive when I rebooted the server. This is expected.

> but I think it might be better to turn the server on first.

Doesn't this approach create a problem when a client is dependent on all servers it is subscribed to? I am not talking about the operation of the client, of course, it is not functioning right anymore. But it will not even start if one of the servers is not running.

> def add_callback_ignore_timeout(sub, func):
>      try:
>          sub.add_callback(func)
>      except caproto.CaprotoTimeoutError:
>         pass

This will help to catch the error but I will want to subscribe as soon as the server is up and running. So I will need to submit this in a separate thread and run it until it is error-free. This is a solution.

**Comment by danielballan** [comment]
Good point. Basically we want a way to throw the subscription into the same cache that would be used to _re_-connect in the event of a dropped connection. That seems like a reasonable thing to add to caproto.

**Comment by matthew-d-jones** [comment]
I also didn't expect this behaviour, I guess because it differs from other EPICS libraries. Documentation is excellent so it was easy to find the reasoning and that it is intentional though :+1: 

A way to monitor when the PV becomes available, without having to make an indefinitely blocking call to subscribe, would be very useful to me.

**Comment by danielballan** [comment]
Thanks @matthew-d-jones. I have been convinced that we should change caproto as suggested.  I think implementing this will actually simplify the code, which is a good sign, but it will require some care. It's roughly at the top of my queue of caproto tasks, though of course if anyone else is interested in taking a crack at it before then I'm happy to support them. :- )

**Comment by danielballan** [comment]
Fix proposed in #608
